aips/aip-x.md

```diff
@@ -0,0 +1,247 @@
+- Flashloan at this rate or lower
+- Variable-rate borrow position with Z APR or lower
+- APR between amount X and Y
+
+By declaring the intent to be trading FA that you own into a NFT that performs those certain operations. This could even bring up a entirely new sets of credit agencies to rate the credibility of each liquidity provider given those certain lending/borrowing NFTs.
+
+All those benefits can help us build a more composable DeFi ecosystem. Users no longer need to specify which protocol they would like to use. This would also incentize for a solution market to find out the right trading solution. Trading, borrowing and lending could be unified under this one framework of offering and claiming process.
+
+## Alternative Solutions
+
+
+## Specification and Implementation Details
+
+We see intent as a programmable lock for on chain assets. Meaning:
+- One party can declare:
+  - Type of Asset they would like to trade in
+  - Condition when this asset could be released
+- The framework provides mechanism to release the lock when such condition is met.
+
+Logic here can then be translated into a Move struct that looks like the following:
+```
+struct TradeIntent<Source, Args> has key {
+    dispensable_resource: Source,
+    argument: Args,
+    self_delete_ref: DeleteRef,
+    expiry_time: u64,
+    witness_type: TypeInfo,
+}
+```
+
+The `TradeIntent` is parameterized by two type parameters:
+1. `Source`: Value that intent issuer is willing to give away
+2. `Argument`: Intent user will use `Argument` as auxillary info to determine whether the unlock condition is met.
+
+In the `TradeIntent`, we also register for a witness type where the releasing `Source` can only be done when such witness is provided.
+
+The lifecycle of an intent will look like the following:
+1. User sends a transaction to the blockchain to create the trading intent by specifying the resource they can offer and the expected condition when such resource can be unlocked.
+2. The intent would be broadcasted using an event.
+3. Anyone can execute this intent as long as the `Witness` can be provided.
+    - We will expect intent solvers to actively monitor the available intents on chain and create transactions to claim the intent.
+
+With those types being generic, we can implement different trading intents easily:
+- If `Target` is instantiated with `FungibleAsset`, it means we are trading a Move value for a specific fungible asset.
+- If `Target` is instantiated with `LinearTransferRef`, it means we are trading a Move value for an Object.
+- If `Source` is instantiated with `TransferRef`, it means we are giving out ownership to an Object given certain condition.
+- If `Source` is instantiated with `FungibleAsset`, it means we are giving out a certain fungible asset given certain condition.
+   - This doesn't quite work because `FungibleAsset` is not storable. The real implementation would look like giving out ownership of a `FungibleStore`
```

Comment:
could this be hidden in the intent? i.e. it stores FungibleStore, but returns extracted FungibleAsset to the caller?

Reply:
Yes it is currently hidden for implementation.